File: cc/app/acc-serv/internal/proto/dns/udp.cc
//
// weproxy@foxmail.com 2022/10/03
//

#include "dns.h"
#include "gx/encoding/binary/binary.h"
#include "gx/net/net.h"
#include "gx/time/time.h"
#include "logx/logx.h"
#include "nx/dns/dns.h"
#include "nx/netio/netio.h"
#include "nx/stats/stats.h"

namespace internal {
namespace proto {
namespace dns {
using namespace nx;

////////////////////////////////////////////////////////////////////////////////
// key_t ...
typedef uint64 key_t;

// makeKey ...
static key_t makeKey(net::IP ip, uint16 port) {
    uint64 a = (uint64)binary::LittleEndian.Uint32(ip.B);  // ipv4
    uint64 b = (uint64)port;
    return a << 16 | b;
}
static key_t makeKey(net::Addr addr) { return makeKey(addr->IP, addr->Port); }

////////////////////////////////////////////////////////////////////////////////
// udpSess_t ...
//
// <NAT-Open Notice> maybe one source client send to multi-target server, likes:
//   caddr=1.2.3.4:100 --> raddr=8.8.8.8:53
//                     --> raddr=4.4.4.4:53
//
struct udpSess_t : public std::enable_shared_from_this<udpSess_t> {
    net::PacketConn ln_;  // to source client, it is net::PacketConn
    net::PacketConn rc_;  // to target server, it is udpConn_t
    net::Addr caddr_;     // source client addr
    Ref<stats::Stats> sta_;
    func<void()> afterClosedFn_;

    udpSess_t(net::PacketConn ln, net::PacketConn rc, net::Addr caddr) : ln_(ln), rc_(rc), caddr_(caddr) {}
    ~udpSess_t() { Close(); }

    // Close ...
    error Close() {
        if (rc_) {
            rc_->Close();
            sta_->Done("closed");
            rc_ = nil;
            if (afterClosedFn_) {
                afterClosedFn_();
            }
        }
        return nil;
    }

    // Start ...
    void Start(const string& target) {
        auto tag = GX_SS(TAG << " UDP_" << nx::NewID() << " " << caddr_ << "->" << target);
        auto sta = stats::NewUDPStats(stats::Type::DNS, tag);

        sta->Start("started");
        sta_ = sta;

        // loopRecvRC
        auto thiz = shared_from_this();
        gx::go([thiz] {
            DEFER(thiz->Close());

            netio::CopyOption opt(time::Second * 10, 0);
            opt.WriteAddr = thiz->caddr_;
            opt.MaxTimes = 1;  // read 1 packet then close
            opt.ReadTimeout = time::Second * 5;
            opt.CopingFn = [thiz](int size) { thiz->sta_->AddSent(size); };

            // copy to ln_ from rc_
            netio::Copy(thiz->ln_, thiz->rc_, opt);
        });
    }

    // WriteToRC ...
    error WriteToRC(const bytez<> buf, net::Addr raddr) {
        if (!rc_) {
            return net::ErrClosed;
        }

        // buf is from source client
        sta_->AddRecv(len(buf));

        AUTO_R(_, err, rc_->WriteTo(buf, raddr));

        return err;
    }
};

// udpSessMap ...
using udpSessMap = Map<key_t, Ref<udpSess_t>>;

////////////////////////////////////////////////////////////////////////////////
// udpConn_t
// to target server
struct udpConn_t : public net::xx::packetConnWrap_t {
    udpConn_t(net::PacketConn pc) : net::xx::packetConnWrap_t(pc) {}

    // ReadFrom ...
    // readFrom target server, and writeTo source client
    virtual R<int, net::Addr, error> ReadFrom(bytez<> buf) override {
        // readFrom target server
        AUTO_R(n, addr, err, wrap_->ReadFrom(buf));
        if (err) {
            return {n, addr, err};
        }

        // cache store
        nx::dns::OnResponse(buf(0, n));

        return {n, addr, nil};
    }

    // WriteTo ...
    // data from source client writeTo target server
    virtual R<int, error> WriteTo(const bytez<> buf, net::Addr raddr) override {
        // TODO ... unpack DNS query packet

        AUTO_R(n, err, wrap_->WriteTo(buf, raddr));

        return {n, err};
    }

    // wrap ...
    static net::PacketConn wrap(net::PacketConn pc) {
        //
        return NewRef<udpConn_t>(pc);
    }
};

////////////////////////////////////////////////////////////////////////////////
// runServLoop ...
error runServLoop(net::PacketConn ln) {
    // sessMap
    auto sessMap = NewRef<udpSessMap>();
    DEFER({
        for (auto& kv : *sessMap) {
            kv.second->Close();
        }
    });

    DEFER(ln->Close());

    bytez<> buf = make(1024 * 4);

    for (;;) {
        // read
        AUTO_R(n, caddr, err, ln->ReadFrom(buf));
        if (err) {
            if (err != net::ErrClosed) {
                LOGS_E(TAG << " err: " << err);
            }
            break;
        } else if (n <= 0) {
            continue;
        }

        LOGS_V(TAG << " ReadFrom() " << caddr);

        // packet data
        auto data = buf(0, n);

        // cache query
        AUTO_R(msg, ans, erx, nx::dns::OnRequest(data));
        if (!erx && ans) {
            // cache answer
            auto b = ans->Bytes();
            if (len(b) > 0) {
                ln->WriteTo(b, caddr);
                continue;
            }
        }

        // lookfor or create sess
        Ref<udpSess_t> sess;

        // use source client addr as key
        // <TODO-Notice>
        //  some user's env has multi-outbound-ip, we will get diff caddr although he use one-same-conn
        key_t key = makeKey(caddr);

        auto it = sessMap->find((key));
        if (it == sessMap->end()) {
            // create a new rc for every new source client
            AUTO_R(c, err, net::ListenPacket(":0"));
            if (err) {
                LOGS_E(TAG << " err: " << err);
                continue;
            }

            // wrap as udpConn_t
            auto rc = udpConn_t::wrap(c);

            // store to sessMap
            sess = NewRef<udpSess_t>(ln, rc, caddr);
            (*sessMap)[key] = sess;

            // remove it after rc closed
            sess->afterClosedFn_ = [sessMap, key] { sessMap->erase(key); };

            auto target = [msg]() -> string {
                if (msg != nil && len(msg->Questions) > 0) {
                    return GX_SS(msg->Questions[0].Name << msg->Questions[0].Type);
                }
                return "";
            }();

            // start recv loop...
            sess->Start(target);
        } else {
            sess = it->second;
        }

        // use 8.8.8.8:53
        static net::Addr raddr = net::MakeAddr(net::IPv4(223, 5, 5, 5), 53);

        // writeTo target server
        err = sess->WriteToRC(data, raddr);
        if (err) {
            return err;
        }
    }

    return nil;
}

}  // namespace dns
}  // namespace proto
}  // namespace internal

File: go/app/acc-serv/internal/proto/dns/udp.go
//
// weproxy@foxmail.com 2022/10/20
//

package dns

import (
	"errors"
	"fmt"
	"net"
	"time"

	"weproxy/acc/libgo/logx"
	"weproxy/acc/libgo/nx"
	"weproxy/acc/libgo/nx/dns"
	"weproxy/acc/libgo/nx/netio"
	"weproxy/acc/libgo/nx/stats"
)

// //////////////////////////////////////////////////////////////////////////////
// udpSess_t ...
//
// <NAT-Open Notice> maybe one source client send to multi-target server, likes:
//
//	caddr=1.2.3.4:100 --> raddr=8.8.8.8:53
//	                  --> raddr=4.4.4.4:53
type udpSess_t struct {
	ln            net.PacketConn // to source client, it is net.PacketConn
	rc            net.PacketConn // to target server, it is udpConn_t
	caddr         net.Addr       // source client addr
	sta           *stats.Stats
	afterClosedFn func()
}

// Close ...
func (m *udpSess_t) Close() error {
	if m.rc != nil {
		m.rc.Close()
		m.sta.Done("closed")
		m.rc = nil
		if m.afterClosedFn != nil {
			m.afterClosedFn()
		}
	}
	return nil
}

// Start ...
func (m *udpSess_t) Start(target string) {
	tag := fmt.Sprintf("%s DNS_%v %v->%s", TAG, nx.NewID(), m.caddr, target)
	sta := stats.NewUDPStats(stats.TypeDNS, tag)

	sta.Start("started")
	m.sta = sta

	// loopRecvRC
	go func() {
		defer m.Close()

		var opt netio.CopyOption
		opt.WriteAddr = m.caddr
		opt.MaxTimes = 1 // read 1 packet then close
		opt.ReadTimeout = time.Second * 5
		opt.CopingFn = func(n int) { m.sta.AddSent(int64(n)) }

		// copy to ln from rc
		netio.CopyPacket(m.ln, m.rc, opt)
	}()
}

// WriteToRC ...
func (m *udpSess_t) WriteToRC(buf []byte, raddr net.Addr) error {
	if m.rc == nil {
		return net.ErrClosed
	}

	// buf is from source client
	m.sta.AddRecv(int64(len(buf)))

	// udpConn_t.WriteTo()
	// unpack data (from source client), and writeTo target server
	_, err := m.rc.WriteTo(buf, raddr)

	return err
}

// udpSessMap ...
type udpSessMap map[nx.Key]*udpSess_t

////////////////////////////////////////////////////////////////////////////////

// udpConn_t
// to target server
type udpConn_t struct {
	net.PacketConn
	// typ socks.AddrType
}

// ReadFrom override
// readFrom target server, will writeTo source client
func (m *udpConn_t) ReadFrom(buf []byte) (n int, addr net.Addr, err error) {
	n, addr, err = m.PacketConn.ReadFrom(buf)
	if err == nil {
		// dns cache store
		dns.OnResponse(buf[:n])
	}
	return
}

// WriteTo override
// data from source client writeTo target server
func (m *udpConn_t) WriteTo(buf []byte, addr net.Addr) (n int, err error) {
	return m.PacketConn.WriteTo(buf, addr)
}

// //////////////////////////////////////////////////////////////////////////////

// runServLoop ...
func runServLoop(ln net.PacketConn) {
	defer ln.Close()

	// sessMap
	sessMap := make(udpSessMap)
	defer func() {
		for _, v := range sessMap {
			v.Close()
		}
	}()

	buf := make([]byte, 1024*8)

	for {
		// read
		n, caddr, err := ln.ReadFrom(buf)
		if err != nil {
			if !errors.Is(err, net.ErrClosed) {
				logx.E("%s err: %v", TAG, err)
			}
			break
		} else if n <= 0 {
			continue
		}

		logx.V("%s ReadFrom() %v", TAG, caddr)

		// packet data
		data := buf[:n]

		// dns cache query
		msg, ans, err := dns.OnRequest(data)
		if err == nil && ans != nil {
			// dns cache answer
			if b := ans.Bytes(); len(b) > 0 {
				ln.WriteTo(b, caddr)
				continue
			}
		}

		// lookfor or create sess
		var sess *udpSess_t

		// use source client addr as key
		// <TODO-Notice>
		//  some user's env has multi-outbound-ip, we will get diff caddr although he use one-same-conn
		key := nx.MakeKey(caddr)

		if v, ok := sessMap[key]; !ok {
			// create a new rc for every new source client
			c, err := net.ListenPacket("udp", ":0")
			if err != nil {
				logx.E("%s err: %v", TAG, err)
				continue
			}

			// wrap as udpConn_t
			rc := &udpConn_t{PacketConn: c}

			// store to sessMap
			sess = &udpSess_t{
				ln:    ln,
				rc:    rc,
				caddr: caddr,
			}
			sessMap[key] = sess

			// remove it after rc closed
			sess.afterClosedFn = func() { delete(sessMap, key) }

			target := func() string {
				if msg != nil && len(msg.Questions) > 0 {
					return msg.Questions[0].Name.String() + msg.Questions[0].Type.String()
				}
				return ""
			}()

			// start recv loop...
			sess.Start(target)
		} else {
			sess = v
		}

		// target DNS server
		raddr := &net.UDPAddr{IP: net.IPv4(223, 5, 5, 5), Port: 53}

		// writeTo target server
		err = sess.WriteToRC(data, raddr)
		if err != nil {
			break
		}
	}
}
